fix email markdown

aimane-chnaif opened the pull request:
@mananjadhav @MonilBhavsar will you please review this?

Proposal:
https://github.com/Expensify/App/issues/10516#issuecomment-1235420788

### Fixed Issues
$ https://github.com/Expensify/App/issues/10516

# Tests
1. Open any chat
2. Type an email address using markdown for a hyperlink: eg `[abc@gmail.com](abc@gmail.com)`
3. Verify that message is displayed correctly as [abc@gmail.com](abc@gmail.com)
4. For various tests inside [], please check below 8 example cases and videos I attached for each platform

# QA
1. Open any chat
2. Type an email address using markdown for a hyperlink: eg `[abc@gmail.com](abc@gmail.com)`
3. Verify that message is displayed correctly as [abc@gmail.com](abc@gmail.com)
4. For various tests inside [], please check below 8 example cases and videos I attached for each platform


### Screenshots

```
1. Hi, [](abc@example.com) is my email
2. Hi, [  ](abc@example.com) is my email
3. Hi, [this](abc@example.com) is my email
4. Hi, [ this](abc@example.com) is my email
5. Hi, [         this](abc@example.com) is my email
6. Hi, [   this        ](abc@example.com) is my email
7. Hi, [this one](abc@example.com) is my email
8. Hi, [   this one   ](abc@example.com) is my email
```

This is how Github comment looks like:
1. Hi, [](abc@example.com) is my email
2. Hi, [  ](abc@example.com) is my email
3. Hi, [this](abc@example.com) is my email
4. Hi, [ this](abc@example.com) is my email
5. Hi, [         this](abc@example.com) is my email
6. Hi, [   this        ](abc@example.com) is my email
7. Hi, [this one](abc@example.com) is my email
8. Hi, [   this one   ](abc@example.com) is my email

#### Web
<img width="1404" alt="web" src="https://user-images.githubusercontent.com/96077027/189087437-8c83d9d0-3b0f-4a49-ac5b-5b6744ae0432.png">



#### Mobile Web
<img width="371" alt="mweb" src="https://user-images.githubusercontent.com/96077027/189087633-e6d3d392-15f5-49dc-81d7-2a6d66f187a9.png">



#### Desktop
<img width="820" alt="desktop" src="https://user-images.githubusercontent.com/96077027/189087659-0f7a53e7-d5c9-477f-a2cb-de50a87f4806.png">



#### iOS
<img width="370" alt="ios" src="https://user-images.githubusercontent.com/96077027/189087679-e0a6dd6f-237d-4549-92b4-a5d46a19128b.png">


#### Android
<img width="405" alt="android" src="https://user-images.githubusercontent.com/96077027/189087697-378637d5-7769-432e-b60c-8308ba6155f5.png">

MonilBhavsar reviewed (commented):
Please add automated tests 🙏

mananjadhav commented:
@aimane-chnaif  I am only going to be able to review this by tomorrow, but as @MonilBhavsar mentioned lets add tests for this

MonilBhavsar commented:
> I am not sure why I am not added as a reviewer. Do you have any idea?

I think because it is in expensify-common repo. Can you request a self-review? I'm not able to add you as a reviewer.

MonilBhavsar commented:
@aimane-chnaif could you please try to request a review from Manan?

aimane-chnaif commented:
> Please don't add any extra text in the linked issue Fixed Issues. Just the link is required, https://github.com/Expensify/App/issues/10516, no Expensify/App#10516

I didn't do any extra. Github did it itself

> There should be a PR Checklist on GH issue? Did it not load for you or did you delete? Please retain it and go through it, verify every item from the checklist and mark them off.

There were no checklist on PR template of this lib repo. You can check here: https://github.com/Expensify/expensify-common/blob/main/.github/PULL_REQUEST_TEMPLATE.md

> There should be a CLA acceptance notification on your side, please accept the CLA if not done already.

Already accepted. This is my 2nd work on this lib repo.

mananjadhav commented:
That's okay @aimane-chnaif, I'll review it. Just tag me once you've added the tests.

> I didn't do any extra. Github did it itself

Again please ignore the comment, yeah I missed that its a part of `expensify-common` and hence it is rendering like that.

aimane-chnaif commented:
@mananjadhav @MonilBhavsar I added automated tests.
FYI: I updated regex from original proposal by invalidating [ and ] inside [ ] to fix cases below **(**`[`, `]`, `(`, `)` inside [ ]**)**
They are also added in automated tests.
```
[(Expensify)](concierge@expensify.com)
[Expensify [Test](test@expensify.com) Test](concierge@expensify.com)
[Expensify [Test](concierge@expensify.com)
[Expensify ]Test](concierge@expensify.com)
```

This is how it looks in github comments:
[(Expensify)](concierge@expensify.com)
[Expensify [Test](test@expensify.com) Test](concierge@expensify.com)
[Expensify [Test](concierge@expensify.com)
[Expensify ]Test](concierge@expensify.com)

And this PR looks exactly same as github.

mananjadhav commented:
Thanks @aimane-chnaif. Please allow me a day to review and test the PR

MonilBhavsar commented:
@mananjadhav could you please review this when you get chance

mananjadhav commented:
Hey thanks for bump @MonilBhavsar. This slipped through the cracks. I'll try to get done today or worst tomorrow. Sorry for the delay!

mananjadhav commented:
Thanks for the patience here @aimane-chnaif. Sorry it took a bit longer to test me. I feel most cases work fine. But the following fails: `[any text](invalid-email-with-space @ xyc.com)` shouldn't turn into a hyperlink. The older regex works fine for this case, but with the current PR this test case fails.

mananjadhav commented:
Hi @aimane-chnaif you're right I am not able to reproduce it, I tried including the email in the display text. `[email@test.com](invalid-email-with-space @ xyc.com)`

This is looking fine. Tests well.

@MonilBhavsar All yours.

🎀 👀 🎀 
C+ reviewed


- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] Any functional components have the `displayName` property
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

mananjadhav reviewed (approved):
LGTM

MonilBhavsar commented on lib/ExpensiMark.js:
```diff
@@ -61,8 +61,8 @@ export default class ExpensiMark {
-                regex: /\[([\w\\s\\d!?&#;]+)\]\(([a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z]+?(\.[a-zA-Z]+)+)\)/gim,
-                replacement: '<a href="mailto:$2">$1</a>',
+                regex: /\[([^[\]]*)\]\(([a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z]+?(\.[a-zA-Z]+)+)\)/gim,
+                replacement: (match, g1, g2) => `<a href="mailto:${g2}">${g1.trim()}</a>`,
```
Can we update this replacement function to something like - 
`g1.trim() ? '<a href="mailto:${g2}">${g1.trim()}</a>' : ''`
Thoughts? cc @mananjadhav

mananjadhav replied:
Agreed with @aimane-chnaif

MonilBhavsar reviewed (commented):
Thank you for adding tests! 🙌

mananjadhav commented:
@MonilBhavsar Are we good to merge? Because we might need another PR for the Expensify App once this is merged.

MonilBhavsar reviewed (approved):
Looks good and works well! 
@nkuoch all yours!

mananjadhav commented:
Yes that is correct, we need to create another PR with the updated commit-hash, and here we'll add screenshots/screencasts of all the platforms  the email markdown renders.

MonilBhavsar commented:
PR to update version in App repo is here https://github.com/Expensify/App/pull/11468

MonilBhavsar commented:
Closed my PR as latest merge commit was referenced and merged before my PR.
